### DISCUSSION TITLE
IE等《インターネット・ブラウザのキャッシュに相当する機能》がFFFTPのダウンロードには？

### DISCUSSION THREAD

**Discussion by Power-Engine** [ffftp/ffftp#276] [Q&A] [answered]
はじめまして
FFFTPのホストからファイルやフォルダをダウンロードする機能に関し、以下の1点をご教示賜りたく、
先ずはご確認いただけると幸いです。

Q1.　エクスプローラーをFTPクライアントとして、同名フォルダ・同名ファイルを複数回ダウンロードする場合、
《一度アクセスし読み込んだファイル内容をOS内で一時保管し次回ダウンロード時に再利用する》こと（キャッシュ
機能と呼ばれるでしょうか。）を有効にできるものと理解しています。
　このような機能に相当する事象が、FFFTPでダウンロードする場合にも有りえるかどうか、知りたいです。

　実際、エクスプローラーをFTPクライアントとして使用しホストからコピーしたファイルで、上記キャッシュ機能の
結果といえる事象が生じることを、自身の経験で知りました。原因は、IEの設定状況で生じる事象と同根と理解します。
→参考　「Web サーバー側でコンテンツを更新しても IE 上に反映されない (キャッシュのお話)」
　https://jpdsi.github.io/blog/internet-explorer-microsoft-edge/IE-Cache/

　そこで知りたいのは、FFFTPを使用しホストからダウンロードしたファイルでも、相当の事象が生じえるかどうか。
　具体的には、複数回ダウンロードするうちに、サーバー側でコンテンツの更新が生じたとき、即時にダウンロード
されるファイル《内容》に反映される、といえるかどうかです。
　そして生じえるならば、その発生有無を決める条件は何か、知りたいです。

以上、ご教示のほど宜しくお願いします。

**Comment by sayurin** [comment, marked as answer]
ffftpにキャッシュ機能は存在しません。操作を行った時点での最新のファイルがダウンロードされます。

**Reply by Power-Engine** [reply]
ご教示ありがとうございました。
FFFTPにキャッシュ機能は存在しない、それも初リリースから最新版に至るまで存在しない、と理解します。
かねてより私は、ウェブ・ブラウザのキャッシュに相当する機能は、FTPクライアントの一般的使われ方から
は無用の長物であり、エクスプローラーのFTPの場合がIEとの関連から特別なのだろう、と想像していました。
エクスプローラーのFTPの使用では、キャッシュ機能無効化や、必ずしもリアルタイムに表示が更新されない
サーバー側のディレクトリ情報（これはFFFTPも？）の更新を、意図的に行う必要があることに留意したいと
思います。

なお改版履歴には「キャッシュ」の文言が次のように登場しますが、それは本題の意味のキャッシュではなく、
《ディレクトリ情報をファイル化すること》を指していると理解します。：
■Ver 1.98
【中略】
・ファイル一覧のキャッシュは既定で無効化されました（以前のバージョンの
　設定が残っている場合はその設定が適用されます）。

ーー
頂いたご助言・ご協力は、私の現在進行中の取り組みに大いに資するところです。
以下ご参考まで、あらためて本題提起の背景となった、私の取り組み中の問題事象について述べます。
最後に〔本題提起により確認できたこと、今後の対応〕でまとめます。

【→6/21 問題事象への取り組みが完結し、以下の記載に誤りが含まれると判明したため削除しました。】

以上／ご協力に感謝申し上げます。

**Reply by sayurin** [reply]
説明が冗長で何を言っているのか頭に入ってきませんが…

ファイルの受け渡しの際にタイムスタンプが変更されていないことは確認済みでしょうか？

> タイムスタンプは更新されているのに内容が旧い

ffftpはキャッシュを行っていないため、ダウンロードされた以上、その時点でサーバーが送信してきたデータです。ファイル内容が古いとしても、サーバーがそのようなデータを送信してきたということになります。

FTPサーバー側のファイルシステム破損なども考慮されるとよいかと。

**Reply by Power-Engine** [reply]
ご指摘賜りありがとうございます。読みづらい長文で失礼しました。若干ながら再編集致しました。
ご助言ありがとうございます。：

> FTPサーバー側のファイルシステム破損なども考慮されるとよいかと。

はい、機器側のファイルシステム破損の可能性から確かめるべき状況と考えます。
これまでの調査では機器やPCのファイルシステムは健全であることを前提に取り組みましたが、
今後、特に機器のファイルシステム破損の可能性も視野に入れます。

> ファイルの受け渡しの際にタイムスタンプが変更されていないことは確認済みでしょうか？

これは次の記述へのご質問でしょう。回答致します。：
> [2] 4月上旬某日、…
> そのようにして回収され、私がメールで受信した当該ファイルを確認した…
> 

　はい、メールの送受信ではファイルのタイムスタンプが保持されています。
　（∵保持されるような圧縮をかけて添付されています。）

ーー
以下ついでながら、

> 5月下旬某日、問題事象の再現トライ（同型の別機体、FFFTP 4.7）で「FFFTPでダウンロード」

を行う中で得た、FFFTPに関する知見を述べます。もし誤りがありましたらご指摘賜れれば幸いです。

先ず、その再現トライでの操作・確認内容です。結果は（６）項のとおりです。：
（１）機器と接続したFFFTP（ホスト側のファイル一覧）で、注目するファイルのタイムスタンプを確認。
（２）機器でそのファイルを上書き更新させる。
（３）FFFTP（ホスト側のファイル一覧）でそのタイムスタンプが、未だ上書き更新されていないのを確認。
（４）そのファイルをFFFTPでローカル側（PC）へダウンロードする。
（５）FFFTP（ローカル側のファイル一覧）でそのファイルのタイムスタンプを見ると、旧いままなのを確認。
　　　（エクスプローラー上で見ても旧いままなのを確認）
（６）そのファイルの内容を確認すると、機器で上書き更新された結果である。即ち問題事象が再現せず。

つまり、私の取り組み中の問題事象：

> タイムスタンプは更新されているのに内容が旧い

に対して、この再現トライでは、問題事象が再現せず、真逆な現象が見られました。：

　タイムスタンプは旧いのに内容が新しい（更新されている）

この現象は「見かけのディレクトリ情報」がダウンロード結果に反映される結果であるため、適宜手動の
［表示を更新］ON操作を実行して最新のディレクトリ情報にすることで回避可、と理解します。
ここで「見かけのディレクトリ情報」とは、FFFTPの使用経験で得た次のような知見に基づく表現です。：
　FFFTPの「ホスト側のファイル一覧」では、「ローカル側のファイル一覧」と異なり、
　ホスト側でファイルが上書き更新されても、タイムスタンプは自動的に更新されない。

実際、上記（３）に以下の（３A）を追加して行ったところ、（５）は以下の（５´）のように変わりました。
（３）FFFTP（ホスト側のファイル一覧）でそのタイムスタンプが、未だ上書き更新されていないのを確認。
（３A）FFFTPの「表示を更新」ボタンONで、そのタイムスタンプが更新されたのを確認。
（４）そのファイルをFFFTPでローカル側（PC）へダウンロードする。
（５´）FFFTP（ローカル側のファイル一覧）でそのファイルのタイムスタンプを見ると、（３A）と同じを確認。

以上／ご理解とご協力に感謝申し上げます。
